File: ml-benchmark/synduce/constraints/even_tree/parity_of_max.f
Inductive Tree = elt Int | node {Int, Tree, Tree};

mod2 = \x: Int. - x (* (/ x 2) 2);

is_even = fix (
  \f: Tree -> Bool. \t: Tree.
  match t with
    elt a -> == (mod2 a) 0
  | node {a, l, r} -> and (and (== (mod2 a) 0) (f l)) (f r)
  end
);

max = \a: Int. \b: Int. if < a b then b else a;

spec = \t: Tree. mod2 ((fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    elt a -> a
  | node {a, l, r} -> max a (max (f l) (f r))
  end
)) t);

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    elt a -> elt a
  | node {a, l, r} -> node {a, l, r} /*Avoid recursions*/
  end
);

main = \t: Tree. if is_even t then spec (target t) else 0;

File: ml-benchmark/synduce/constraints/even_tree/parity_of_max.ml
type tree = Elt of int | Node of int * tree * tree

let mod2 x = x - ((x / 2) * 2)

let rec is_even t =
  match t with
  | Elt a -> mod2 a == 0
  | Node (a, l, r) -> mod2 a == 0 && is_even l && is_even r

let max a b = if a < b then b else a

let spec t =
  let rec f t =
    match t with
    | Elt a -> a
    | Node (a, l, r) -> max a (max (f l) (f r))
  in
  mod2 (f t)

val target: tree -> tree compress
let target t =
  match t with
  | Elt a -> Elt a
  | Node (a, l, r) -> Node (a, l, r)

let program t = if is_even t then spec (target t) else 0